sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,317 @@
+              << SPI_DEVICE_RXF_ADDR_BASE_OFFSET;
+  rxf_addr |= (SPI_DEVICE_RXF_END & SPI_DEVICE_RXF_ADDR_LIMIT_MASK)
+              << SPI_DEVICE_RXF_ADDR_LIMIT_OFFSET;
+  mmio_region_write32(config.base_addr, SPI_DEVICE_RXF_ADDR_REG_OFFSET,
+                      rxf_addr);
+
+  // Set up the range of the TXF region, inclusive.
+  uint32_t txf_addr = 0;
+  txf_addr |= (SPI_DEVICE_TXF_BASE & SPI_DEVICE_TXF_ADDR_BASE_MASK)
+              << SPI_DEVICE_TXF_ADDR_BASE_OFFSET;
+  txf_addr |= (SPI_DEVICE_TXF_END & SPI_DEVICE_TXF_ADDR_LIMIT_MASK)
+              << SPI_DEVICE_TXF_ADDR_LIMIT_OFFSET;
+  mmio_region_write32(config.base_addr, SPI_DEVICE_TXF_ADDR_REG_OFFSET,
+                      txf_addr);
+
+  spi_out->base_addr = config.base_addr;
+
+  return kDifSpiDeviceInitOk;
+}
+
+/**
+ * Calculates the *depth*, or bytes-in-use of a SPI FIFO (which is implemented
+ * as a ring buffer).
+ *
+ * If the read/write pointers are in-phase, this is just their difference; if
+ * they're out-of-phase, one of the pointers has "wrapped around", so there is
+ * an additional FIFO's worth of data pending.
+ */
+static size_t fifo_depth(uint32_t write_addr, uint32_t read_addr,
+                                size_t fifo_len) {
+  uint32_t write_phase = write_addr & PHASE_MASK;
+  uint32_t read_phase = read_addr & PHASE_MASK;
+  uint32_t write_ptr_val = write_addr & VALUE_MASK;
+  uint32_t read_ptr_val = read_addr & VALUE_MASK;
+
+  // This represents the case where the valid data of the fifo is "inclusive",
+  // i.e., the buffer looks like (where a / represents valid data):
+  // [ /////       ]
+  //   ^    ^
+  //   read write
+  if (write_phase == read_phase) {
+    return write_ptr_val - read_ptr_val;
+  }
+
+  // This represents the case where the valid data of the fifo is "exclusive",
+  // i.e., the buffer looks like (where a / represents valid data):
+  // [/      //////]
+  //   ^     ^
+  //   write read
+  return fifo_len - (read_ptr_val - write_ptr_val);
+}
+
+/**
+ * Increments the FIFO address pointed to by |addr| by |increment|.
+ *
+ * If the pointer would wrap around the FIFO buffer, it gets wrapped and the
+ * phase bit is flipped, since this causes the FIFO to switch between
+ * "inclusive" and "exclusive" modes (see |fifo_depth()| above).
+ *
+ * A fifo pointer can be approximated by a 12-bit sign-and-magnitude integer.
+ */
+static void fifo_increment(uint32_t *addr, uint32_t increment,
+                                  size_t fifo_len) {
+  uint32_t phase_bit = *addr & PHASE_MASK;
+  uint32_t inc_ptr = (*addr & VALUE_MASK) + increment;
+
+  // If we would overflow, wrap and flip the sign bit.
+  if (inc_ptr >= fifo_len) {
+    inc_ptr = (inc_ptr - fifo_len) & VALUE_MASK;
+    phase_bit = ~phase_bit & PHASE_MASK;
+  }
+
+  *addr = inc_ptr | phase_bit;
+}
+
+#define IS_WORD_ALIGNED(addr) (((uintptr_t)(addr)&0b11) == 0)
+
+dif_spi_device_send_result_t dif_spi_device_send(const dif_spi_device_t *spi,
```

Comment:
I think somewhere in this function and its receive counterpart, we must state that the SRAM does **not** support byte access and all that code related to alignment is not just an optional optimization.